includes/chart.js

```diff
@@ -0,0 +1,330 @@
+-----------------------------------------------------------------------------
+
+// NOTES
+
+The chart.js contains all the functions to render the svg flowchart based on data uploaded.
+The operations executed here are:
+    1. Generate the nodes and links arrays
+    2. Create the nodes and set all the required attributes
+    3. Generata the links
+
+*/
+
+// Define nodes following the template
+const row_template = {
+    nodes: [
+        {
+            id: 1, x: 0, y: 30, color: '#ffc70f', shape: 'polygon', type: 'activationCondition',
+            points: '30,60 9,30 30,0 90,0 111,30 90,60',
+        },
+        {
+            id: 2, x: 120, y: 30, color: '#0fafff', shape: 'circle', type: 'attribute',
+            size: '5px', radiusX: 48, radiusY: 36,
+        },
+        {
+            id: 3, x: 240, y: 0, color: '#ffc70f', shape: 'rect', type: 'aim',
+            width: 100, height: 60,
+        },
+        {
+            id: 4, x: 420, y: 0, color: '#0fafff', shape: 'rect', type: 'directObject',
+            width: 110, height: 60, rx: 30, ry: 30,  // rx/ry for rounded corners
+        },
+        {
+            id: 5, x: 240, y: 90, color: '#0fafff', shape: 'rect', type: 'executionConstraint',
+            width: 110, height: 60, rx: 30, ry: 30,  // rx/ry for rounded corners
+        },
+        {
+            id: 6, x: 420, y: 90, color: '#0fafff', shape: 'rect', type: 'indirectObject',
+            width: 110, height: 60, rx: 30, ry: 30,  // rx/ry for rounded corners
+        },
+    ],
+    links: [
+        { source: 1, target: 2 },
+        { source: 2, target: 3 },  // include deontic info here
+        { source: 3, target: 4 },
+        { source: 5, target: 3 },
+        { source: 4, target: 6 },
+    ],
+};
+
+const colors = { formal: '#99cc00', informal: '#f4b183' }
+
+const rowAttributes = [
+    "id",
+    "statementType",
+    "attribute",
+    "deontic",
+    "aim",
+    "directObject",
+    "typeOfDirectObject",
+    "indirectObject",
+    "typeOfIndirectObject",
+    "activationCondition",
+    "executionConstraint",
+    "orElse"
+];
+
+function addNodesAndLinks(rowValues) {
+    var svgContainer = document.getElementById("svgContainer");
+    var svg = svgContainer.querySelector("svg");
+
+    const rowHeight = 180;
+    const rowX = 75, rowY = 30;
+
+    // Add nodes
+    rowValues.forEach(function (row, i) {
+        entries = rowAttributes.map((attribute, idx) => {
+            return [attribute, row[idx]];
+        });
+        row = Object.fromEntries(entries);
+        row_template.nodes.forEach(function (template_node) {
+
+            if (!row[template_node.type]) {
+                return;
+            }
+
+            var group = document.createElementNS("http://www.w3.org/2000/svg", "g");
+            var newNode;
+            var x = template_node.x + rowX;
+            var y = template_node.y + rowY + (i * rowHeight);
+            if (template_node.shape === 'polygon') {
```

Comment:
There are quite some if/else toggles for the different node shapes. I could imagine it to be more extensible and performant if these happen once (or at least less often). I would need to think about it (or maybe co-work for a bit with you) to see how that can be simplified.

The sniff check says it's quite some code repetition that could probably be avoided.

Reply:
All common attributes are assigned outside of the if-else ifs, the code blocks are the slightly different actions associated with each shape. So as far as I can see, there is nothing that can be explicitly de-duplicated...

I've made a separate issue to revisit this (#24) to see if there's any viable refactoring to be done, that way we can at least close this PR and move on with smaller changes from now on.